sqlserver/datadog_checks/sqlserver/schemas.py

```diff
@@ -0,0 +1,365 @@
+            "agent_version": datadog_agent.get_version(),
+            "dbms": "sqlserver",
+            "kind": "sqlserver_databases",
+            "collection_interval": schemas_collection_interval,
+            "dbms_version": None,
+            "tags": self._check.non_internal_tags,
+            "cloud_metadata": self._check._config.cloud_metadata,
+        }
+        self._dataSubmitter = SubmitData(self._check.database_monitoring_metadata, base_event, self._log)
+
+    """Collects database information and schemas and submits to the agent's queue as dictionaries
+    schema dict
+    key/value:
+        "name": str
+        "id": str
+        "owner_name": str
+        "tables" : list of tables dicts
+            table
+            key/value:
+                "id" : str
+                "name" : str
+                columns: list of columns dicts
+                    columns
+                    key/value:
+                        "name": str
+                        "data_type": str
+                        "default": str
+                        "nullable": bool
+            indexes : list of index dicts
+                index
+                key/value:
+                    "name": str
+                    "type": str
+                    "is_unique": bool
+                    "is_primary_key": bool
+                    "is_unique_constraint": bool
+                    "is_disabled": bool,
+                    "column_names": str
+            foreign_keys : list of foreign key dicts
+                foreign_key
+                key/value:
+                    "foreign_key_name": str
+                    "referencing_table": str
+                    "referencing_column": str
+                    "referenced_table": str
+                    "referenced_column": str
+            partitions: partition dict
+                partition
+                key/value:
+                    "partition_count": int
+    """
+
+    @tracked_method(agent_check_getter=agent_check_getter)
+    def collect_schemas_data(self):
+        self._dataSubmitter.reset()
+        self._dataSubmitter.set_base_event_data(
+            self._check.resolved_hostname,
+            self._check.non_internal_tags,
+            self._check._config.cloud_metadata,
+            "{},{}".format(
+                self._check.static_info_cache.get(STATIC_INFO_VERSION, ""),
+                self._check.static_info_cache.get(STATIC_INFO_ENGINE_EDITION, ""),
+            ),
+        )
+
+        # returns if to stop, True means stop iterating.
+        def fetch_schema_data(cursor, db_name):
+            db_info = self._query_db_information(db_name, cursor)
+            schemas = self._query_schema_information(cursor)
+            self._dataSubmitter.store_db_info(db_name, db_info)
+            for schema in schemas:
+                tables = self._get_tables(schema, cursor)
+                tables_chunks = list(get_list_chunks(tables, self.TABLES_CHUNK_SIZE))
```

Comment:
Instead of fetching all tables into memory then dividing them by chunks, can we try use cursor.fetchmany to fetch rows by chunk size?

Reply:
yeah, so its a bit of an MVP thing. 
Chunks allow to minimise amount of RAM and probably are good in case of a corrupted payload(if applicable) - at least one request is not getting through. 

But for performance wise 500 chunks were ~15% better on a local host.
so why then 50  ? Well, when working on Postgres Seth encountered the issue that if integration is to fast to accumulate data then queue cannot handle it. So 50 is like  to artificially slow down intake a bit. We could just have a time limit - if integration collects data faster than in X seconds than we put a sleep before it can submit to the queue ... 

In that sense slowing down integration by cursor.fetchmany isn't more advantageous than doing it by chunks in sql queries. The best would be to know the state of the queue but we cant as we've discussed in our meeting. 
Here is the postgres code - 

https://github.com/DataDog/integrations-core/blob/3b512ede800360a51c7347f9c9c9ecf6d50d5d69/postgres/datadog_checks/postgres/metadata.py#L285